lib/webdriveragent.js

```diff
@@ -244,215 +239,94 @@ class WebDriverAgent {
-        let msg = `Unable to start WebDriverAgent: ${err}`;
-        log.error(msg);
-        return reject(msg);
-      }
+      })();
     });
   }
 
   async waitForStart (startTime) {
-    // we have to wait for the sim to start before we can tail the log file
-    if (!this.realDevice) {
-      await systemLogExists(this.device);
-    }
-
-    let agentUrl;
-    let lineCount = 0;
-    let showWaitingMessage = true; // turn off logging once we have hit the end
-
-    let startDetector = (stdout) => {
-      let match = AGENT_STARTED_REGEX.exec(stdout);
-      if (match) {
-        if (this.realDevice) {
-          // on a real device there may already be system logs that need to be
-          // passed before we get to the real startup logs, otherwise
-          // it will be "done" before booting
-          // Expect a line like:
-          //     Dec  7 10:55:32 iamPhone XCTRunner(WebDriverAgentLib)[386] <Notice>: ServerURLHere->http://localhost:8100<-ServerURLHere
-          // need to get the date and add the year from the start date
-          let dateString = `${stdout.substr(0, 6)} ${startTime.getFullYear()} ${stdout.substr(7, 8)}`;
-          let buildTime = new Date(dateString);
-          if (!buildTime.isAfter(startTime)) {
-            return false;
-          }
+    // try to connect once every 0.5 seconds, until `wdaLaunchTimeout` is up
+    log.debug(`Waiting up to ${this.wdaLaunchTimeout}ms for WebDriverAgent to start`);
+    try {
+      let retries = parseInt(this.wdaLaunchTimeout / 500, 10);
+      await retryInterval(retries, 500, async () => {
+        if (this.xcodebuild.processExited) {
+          // there has been an error elsewhere and we need to short-circuit
+          return;
         }
-
-        agentUrl = match[1];
-        if (!agentUrl) {
-          log.errorAndThrow(new Error('No url detected from WebDriverAgent'));
+        try {
+          let res = await request(`${this.url.href}status`);
+          res = JSON.parse(res);
+          if (res.status !== 0) {
+            throw new Error(`Received non-zero status code from WDA server: '${res.status}'`);
+          }
+        } catch (err) {
+          throw new Error(`Unable to connect to running WebDriverAgent: ${err.message}`);
         }
-        log.info(`Detected that WebDriverAgent is running at url '${agentUrl}'`);
-        showWaitingMessage = false;
-        return true;
-      }
-
-      // periodically log, so it does not look like everything died
-      lineCount++;
-      let threshold = this.realDevice ? 5000 : 200;
-      if (showWaitingMessage && lineCount % threshold === 0) {
-        log.debug('Waiting for WebDriverAgent server to finish loading...');
-      }
-
-      return false;
-    };
-
-    log.info('Waiting for WebDriverAgent to start on device');
-    await this.deviceLogs.start(startDetector);
-    log.info(`WebDriverAgent started at url '${agentUrl}'`);
+      });
+      let startupTime = Date.now() - startTime;
```

Comment:
shortly said - process.hrtime has been initially designed to measure intervals with maximum possible precision. And Date is a general class to represent calendar dates in apps.